### PR TITLE
Setup and run partitions tests on EKS

### PR DESCRIPTION
**Pull request by ndhanushkodi**
Changes proposed in this PR:
- Run partitions tests on EKS
- For cross partition communication, the networking setup needs to be such that the node ips of all clients need to be able to talk with the node ips of all servers and vice versa (traffic should be routeable and allowed). Clients across partitions do not need to be able to route to each other.
- To make this work on EKS, the terraform has been modified to:
  - Set up a 2 clusters in 2 VPCs without overlapping IPs
  - Set up peering between those VPCs
  - Add security group ingress rules to each set of nodes' security group, to allow all TCP traffic from the other VPC
  - Add route table entries such that traffic going to the other VPC is routed through the peer
- A few tweaks to partitions acceptance tests so that they pass
- Outside of this PR, I added permissions to the ci role used for helm tests to be able to do anything on EC2 (ec2*) using the aws cli: 
```
aws iam put-role-policy --role-name service-circleci-consul-helm-tests --policy-name CircleCIManageEC2 --policy-document file://ec2policy.json
```
where ec2policy.json looks like this:
```
{
   "Version": "2012-10-17",
   "Statement": [
    {
      "Effect": "Allow",
      "Action": "ec2:*",
      "Resource": "*"
    }
   ]
}
```

How I've tested this PR:
- Running EKS acceptance tests in the pipeline: https://app.circleci.com/pipelines/github/hashicorp/consul-k8s/3871/workflows/854039f8-5c8c-4b29-8e43-6a45445025c3 (without tproxy)
- EKS acc tests with tproxy passes here: https://app.circleci.com/pipelines/github/hashicorp/consul-k8s/3875/workflows/7b4c2445-4848-44f9-b3e0-01fca56cd9e2/jobs/34507

How I expect reviewers to test this PR:
- Review-- I'd like especially like feedback on:
  - ~This currently relies on always deploying 2 clusters, and isn't conditional on any variables for whether to set up the peering. Since the pipeline always deploys 2 clusters, I thought this could be alright. We could even turn cluster_count from a var to a local, to make it clearer the config is meant for deploying 2 clusters. In a case where someone wants to deploy 1 cluster, they'd need to remove the config from line 92 down, and change the local cluster_count to 1.~ Edited to work with cluster_count = 1 or 2.

Checklist:
- [x] Tests added
- [ ] CHANGELOG entry added 
  > HashiCorp engineers only, community PRs should not add a changelog entry.
  > Entries should use present tense (e.g. Add support for...)

### CODE REVIEW

File: .circleci/config.yml
Comment (ndhanushkodi) [inline .circleci/config.yml:106]: To be removed before merging

File: .circleci/config.yml
Comment (ndhanushkodi) [inline .circleci/config.yml:977]: To be removed before merging

File: acceptance/tests/partitions/partitions_test.go
Comment (ndhanushkodi) [inline acceptance/tests/partitions/partitions_test.go:38]: Should I add a temporary config value for EKS? To enable this only on Kind and EKS until we get all the clouds running?

File: acceptance/tests/partitions/partitions_test.go
Comment (ndhanushkodi) [inline acceptance/tests/partitions/partitions_test.go:261]: In a cloud env, there will be more than one client agent bc there will be multiple nodes. I could add back this check just for Kind but it seemed not essential

File: acceptance/tests/partitions/partitions_test.go
Comment (ndhanushkodi) [inline acceptance/tests/partitions/partitions_test.go:183]: In EKS theres a service called "kubernetes" which points to the API as well, so I removed the `if cfg.UseKind`.

File: acceptance/tests/partitions/partitions_test.go
Comment (thisisnotashwin) [reply acceptance/tests/partitions/partitions_test.go:38]: hmm.. maybe we can explicitly skip AKS and GKE? that way one can peel away those checks as those clouds are configured for these tests? is that possible?

File: acceptance/tests/partitions/partitions_test.go
Comment (thisisnotashwin) [inline acceptance/tests/partitions/partitions_test.go:171]: NICE!

File: acceptance/tests/partitions/partitions_test.go
Comment (thisisnotashwin) [reply acceptance/tests/partitions/partitions_test.go:261]: i agree..it's not essential!!

**Review by thisisnotashwin — APPROVED** [review]
THIS LOOKS GREAT!! EXCELLENTLY DONE NITYA!

File: charts/consul/test/terraform/eks/main.tf
Comment (thisisnotashwin) [inline charts/consul/test/terraform/eks/main.tf:94]: 🙏 thanks for the clarification. I was about to ask about it 😁

Comment (thisisnotashwin) [reply charts/consul/test/terraform/eks/main.tf:94]: @ishustava is this true for AKS as well? i noticed in your PR that is doesn't assume that the cluster_count is 2 and computes it.

Comment (ishustava) [reply charts/consul/test/terraform/eks/main.tf:94]: 🤔 yeah I didn't assume that. I left a comment about it above.

Comment (ndhanushkodi) [reply charts/consul/test/terraform/eks/main.tf:94]: Yeah for AKS, the cluster_count is allowed to be 1 or 2. In the current state of this EKS PR, cluster_count is only allowed to be 2, unless you make some changes to the terraform (like commenting out the whole vpc block). Because these EKS resources are like unidirectional, so I needed to set them up such that ingress rule from cluster 0 --> 1 works, and vice versa, so I couldn't just rely on the count variable for these resources.

File: charts/consul/test/terraform/eks/main.tf
Comment (ishustava) [inline charts/consul/test/terraform/eks/main.tf:146]: I'm curious why do we need a data resource if we have the vpc module created in this terraform. In my experience, I usually use data resource for resources that exist in a different terraform state.

Should this be a local variable instead?

File: charts/consul/test/terraform/eks/main.tf
Comment (ndhanushkodi) [reply charts/consul/test/terraform/eks/main.tf:94]: Yeah for AKS, the cluster_count is allowed to be 1 or 2. In the current state of this EKS PR, cluster_count is only allowed to be 2, unless you make some changes to the terraform (like commenting out the whole vpc block). Because these EKS resources are like unidirectional, so I needed to set them up such that ingress rule from cluster 0 --> 1 works, and vice versa, so I couldn't just rely on the count variable for these resources.

**Review by ishustava — COMMENTED** [review]
Amazing work, Nitya!!

I left some comments inline. Sorry for taking so long to review!

File: charts/consul/test/terraform/eks/main.tf
Comment (ndhanushkodi) [reply charts/consul/test/terraform/eks/main.tf:94]: But it looks like @ishustava 's suggestion for conditionalizing the count may work! terraform apply is chugging along, so 🤞 that destroy works alright too

File: acceptance/tests/partitions/partitions_test.go
Comment (ndhanushkodi) [reply acceptance/tests/partitions/partitions_test.go:38]: should we just consider partitions-eks the feature branch? i saw the aks pr was made against this branch

File: charts/consul/test/terraform/eks/main.tf
Comment (ndhanushkodi) [reply charts/consul/test/terraform/eks/main.tf:93]: Edited with your suggestion! Passed here: https://app.circleci.com/pipelines/github/hashicorp/consul-k8s/3906/workflows/62c03cdd-aaf4-49e5-bb65-3f7335ddef8f/jobs/34923

File: charts/consul/test/terraform/eks/main.tf
Comment (ishustava) [reply charts/consul/test/terraform/eks/main.tf]: I think you can use vpc module output so you can combine public and private route table ids into one array: `[module.vpc[count.index].public_route_table_ids[0], module.vpc[count.index].private_route_table_ids[0]]`

Comment (ishustava) [reply charts/consul/test/terraform/eks/main.tf]: I think the vpc module creates two route tables that are shared between subnets. So there should be one private and one public route table per VPC

File: acceptance/tests/partitions/partitions_test.go
Comment (ndhanushkodi) [reply acceptance/tests/partitions/partitions_test.go:38]: Ya that makes sense, I changed the base to cloud-partitions-tests as the feature branch

**Review by ishustava — APPROVED** [review]
🎉 🎉 

I'm curious what you think about this comment: https://github.com/hashicorp/consul-k8s/pull/953#discussion_r782523964

Otherwise, looks good! I won't hold up my approval on that as I don't think it's blocking

File: charts/consul/test/terraform/eks/main.tf
Comment (thisisnotashwin) [inline charts/consul/test/terraform/eks/main.tf:161]: will the below work? or will it interact poorly with the `aws_route` in the next step? it could potentially work as rtseks[0]
```suggestion
# Get the route table ids that are associated with eks cluster 0, so we can create routes in those route tables.
data "aws_route_tables" "rtseks" {
  count  = var.cluster_count > 1 ? 2 : 0
  vpc_id = module.vpc[count.index].vpc_id

  filter {
    name   = "tag:Name"
    values = [format("%s-*", module.eks[count.index].cluster_id)]
  }
}
```

Comment (ndhanushkodi) [reply charts/consul/test/terraform/eks/main.tf:161]: oh yeah that could possibly work! but I think i'll end up removing these per iryna's suggestion to use the vpc module to grab the route tables instead!

Comment (thisisnotashwin) [reply charts/consul/test/terraform/eks/main.tf:161]: sounds good!! 👍

**Comment by ndhanushkodi** [comment]
Thanks all! Sorry for missing that comment earlier Iryna! I'm testing that suggestion here: https://app.circleci.com/pipelines/github/hashicorp/consul-k8s?branch=partitions-eks&filter=all

File: charts/consul/test/terraform/eks/main.tf
Comment (ndhanushkodi) [reply charts/consul/test/terraform/eks/main.tf]: sorry for missing this earlier! your suggestion worked for me and I was able to remove the route table data block!

**Pull request merged**
